pkg/messaging/mqtt/pubsub_test.go

```diff
@@ -0,0 +1,428 @@
+			Payload:   tc.payload,
+		}
+		err := pubsub.Publish(topic, expectedMsg)
+		assert.Nil(t, err, fmt.Sprintf("%s: got unexpected error: %s\n", tc.desc, err))
+
+		receivedMsg := <-msgChan
+		if tc.payload == nil {
+			assert.Equal(t, 0, len(receivedMsg.Payload), fmt.Sprintf("%s: expected %+v got %+v\n", tc.desc, tc.payload, receivedMsg.Payload))
+		} else {
+			assert.Equal(t, tc.payload, receivedMsg.Payload, fmt.Sprintf("%s: expected %+v got %+v\n", tc.desc, tc.payload, receivedMsg.Payload))
+		}
+	}
+
+	token = client.Unsubscribe(topic)
+	token.Wait()
+	assert.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
+	token = client.Unsubscribe(fmt.Sprintf("%s.%s", topic, subtopic))
+	token.Wait()
+	assert.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
+}
+
+func TestSubscribe(t *testing.T) {
+	// Creating client to Publish messages to subscribed topic.
+	client, err := newClient(address, "clientIDD", 30*time.Second)
+	assert.Nil(t, err, fmt.Sprintf("got unexpected error: %s", err))
+	t.Cleanup(func() {
+		client.Unsubscribe()
+		client.Disconnect(5)
+	})
+
+	cases := []struct {
+		desc     string
+		topic    string
+		clientID string
+		err      error
+		handler  messaging.MessageHandler
+	}{
+		{
+			desc:     "Subscribe to a topic with an ID",
+			topic:    topic,
+			clientID: "clientid1",
+			err:      nil,
+			handler:  handler{false, "clientid1"},
+		},
+		{
+			desc:     "Subscribe to the same topic with a different ID",
+			topic:    topic,
+			clientID: "clientid2",
+			err:      nil,
+			handler:  handler{false, "clientid2"},
+		},
+		{
+			desc:     "Subscribe to an already subscribed topic with an ID",
+			topic:    topic,
+			clientID: "clientid1",
+			err:      nil,
+			handler:  handler{false, "clientid1"},
+		},
+		{
+			desc:     "Subscribe to a topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s", topic, subtopic),
+			clientID: "clientid1",
+			err:      nil,
+			handler:  handler{false, "clientid1"},
+		},
+		{
+			desc:     "Subscribe to an already subscribed topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s", topic, subtopic),
+			clientID: "clientid1",
+			err:      nil,
+			handler:  handler{false, "clientid1"},
+		},
+		{
+			desc:     "Subscribe to an empty topic with an ID",
+			topic:    "",
+			clientID: "clientid1",
+			err:      mqtt_pubsub.ErrEmptyTopic,
+			handler:  handler{false, "clientid1"},
+		},
+		{
+			desc:     "Subscribe to a topic with empty id",
+			topic:    topic,
+			clientID: "",
+			err:      mqtt_pubsub.ErrEmptyID,
+			handler:  handler{false, ""},
+		},
+	}
+	for _, tc := range cases {
+		err = pubsub.Subscribe(tc.clientID, tc.topic, tc.handler)
+		assert.Equal(t, err, tc.err, fmt.Sprintf("%s: expected: %s, but got: %s", tc.desc, err, tc.err))
+
+		if tc.err == nil {
+			token := client.Subscribe(tc.topic, qos, nil)
+			assert.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
```

Comment:
Why do you subscribe here? You already subscribed in the `pubsub.Subscribe`.

Reply:
I can't send messages from client without subscribing to that topic.
I tried client.Connect() first, but that is giving an mqtt error of "Not connected".

Reply:
msgChan is used inside the `Handle()` function, which is a method of `handler`. So, if I make it function-level.
I'll have to pass it to `mqttHandler() function`, which is not exported from package:rabbitmq.
So, this will make all testing really difficult.
I can, perhaps redefine it inside each function. 

`SetDefaultPublishHandler(mqttHandler(handler{false, id}))`: I'll do this part today. And figure out how to complete the test.